cpp/src/arrow/compute/row/row_encoder_internal.h

```diff
@@ -142,41 +159,33 @@ struct ARROW_EXPORT VarLengthKeyEncoder : KeyEncoder {
 
   Status Encode(const ExecValue& data, int64_t batch_length,
                 uint8_t** encoded_bytes) override {
+    auto handle_next_valid_value = [&encoded_bytes](std::string_view bytes) {
+      auto& encoded_ptr = *encoded_bytes++;
+      *encoded_ptr++ = kValidByte;
+      util::SafeStore(encoded_ptr, static_cast<Offset>(bytes.size()));
+      encoded_ptr += sizeof(Offset);
+      memcpy(encoded_ptr, bytes.data(), bytes.size());
+      encoded_ptr += bytes.size();
+    };
+    auto handle_next_null_value = [&encoded_bytes]() {
+      auto& encoded_ptr = *encoded_bytes++;
+      *encoded_ptr++ = kNullByte;
+      util::SafeStore(encoded_ptr, static_cast<Offset>(0));
+      encoded_ptr += sizeof(Offset);
+    };
     if (data.is_array()) {
-      VisitArraySpanInline<T>(
-          data.array,
-          [&](std::string_view bytes) {
-            auto& encoded_ptr = *encoded_bytes++;
-            *encoded_ptr++ = kValidByte;
-            util::SafeStore(encoded_ptr, static_cast<Offset>(bytes.size()));
-            encoded_ptr += sizeof(Offset);
-            memcpy(encoded_ptr, bytes.data(), bytes.size());
-            encoded_ptr += bytes.size();
-          },
-          [&] {
-            auto& encoded_ptr = *encoded_bytes++;
-            *encoded_ptr++ = kNullByte;
-            util::SafeStore(encoded_ptr, static_cast<Offset>(0));
-            encoded_ptr += sizeof(Offset);
-          });
+      VisitArraySpanInline<T>(data.array, handle_next_valid_value,
```

Comment:
```suggestion
      DCHECK_EQ(data.length(), batch_length);
      VisitArraySpanInline<T>(data.array, handle_next_valid_value,
```

Reply:
Just found in this branch `batch_length` is not actually used nor checked - meaning we are assuming `data` to be of length `batch_length`. Better assert it.

Reply:
There are other places of this pattern, e.g., in other encoder sub-classes, and `AddLength` methods. Could you update them as well?

Reply:
no I just found that method is low efficient in compute length...

Reply:
Variable add length might not need that, since if it's scalar, the length is computed and can be accelerate without calling `handle_null()` or handle valie

Reply:
> Variable add length might not need that, since if it's scalar, the length is computed and can be accelerate without calling `handle_null()` or handle valie

I don't understand what you meant. I'm simply suggesting that we assert the the length of `data` array is equal to the length that the caller claims (`batch_length`).